### DISCUSSION TITLE
I am not getting any state from onTransition

### DISCUSSION THREAD

**Discussion by theprantadutta** [statelyai/xstate#1686] [Q&A]
So here's my machine.
```
import { assign, createMachine, interpret } from "xstate";

export interface toggleAuth {
  isAuthenticated: boolean;
  user: {
    name: string | undefined;
  };
}

// console.log(getCachedData());

export const authMachine = createMachine<toggleAuth>({
  id: "auth",
  initial: "unauthenticated",
  context: {
    isAuthenticated: false,
    user: {
      name: undefined,
    },
  },
  states: {
    authenticated: {
      on: {
        toggle: {
          target: "unauthenticated",
          actions: assign((context, event) => {
            return {
              ...context,
              isAuthenticated: true,
              user: { name: "Pranta" },
            };
          }),
        },
      },
    },
    unauthenticated: {
      on: {
        toggle: {
          target: "authenticated",
          actions: assign((context, event) => {
            return {
              ...context,
              isAuthenticated: false,
              user: { name: undefined },
            };
          }),
        },
      },
    },
  },
});

export const authServices = interpret(authMachine).onTransition((state) => {
  console.log("State: ", state.value);
});

authServices.start();
```
Now I am able to toggle the states and context and everything. But the authService is not console-logging the state value like it should every time the state changes. It only runs once. 
I am console-logging these state values in another component like Navbar and it works fine there. But I need to detect the state change to persist the state. So how do I do that inside this service? To be precise, how do I make my service run everytime there is a change in states.
I am using nextjs. So is it a problem with react or am I doing something wrong? Thanks in advance

**Comment by davidkpiano** [comment]
>  It only runs once.

Are you sending events to the auth service? It won't change state unless you do, according to the machine definition.

**Reply by theprantadutta** [reply]
Thanks for replying so fast, I am sending the event like the following. 
```
send('toggle')
```
That actually toggles the auth.

**Reply by theprantadutta** [reply]
I have also tried the following way.
```
const [state, sendService] = useService(authServices);
sendService('toggle')
```
This didn't work

**Reply by davidkpiano** [reply]
Please create a [CodeSandbox (React template)](https://codesandbox.io/s/xstate-react-template-3t2tg) for this. Thank you!

**Reply by theprantadutta** [reply]
Here: https://codesandbox.io/live/z4sv5rd

**Reply by davidkpiano** [reply]
The `toggleMachine` is like a "blueprint" for how an instance of it (a service) should behave.

The service you created is a completely different instance from the service you created inside the component.

This should do what you want:

```jsx
function App() {
  const [current, send] = useMachine(toggleMachine);
  const active = current.matches("active");
  const { count } = current.context;

  useEffect(() => console.log(current), [current]);

  return (
    <div className="App">
      <h1>XState React Template</h1>
      <h2>Fork this template!</h2>
      <button onClick={() => send("TOGGLE")}>
        Click me ({active ? "✅" : "❌"})
      </button>{" "}
      <code>
        Toggled <strong>{count}</strong> times
      </code>
    </div>
  );
}
```

Out of curiosity, where did you get the impression that the `service` you created above refers to _all_ future services? That doesn't make sense 😕

**Reply by theprantadutta** [reply]
You didn't get me. My code was actually working. I did exactly what you did here. But is there a way for me to invoke a service inside toggleMachine (if toggleMachine was in toggleMachine.js)? Let's suppose I want to add a listener to listen for every time the state changes. Now I know I can do this inside useEffect in a component. But I want to use this in multiple components of my code. Now every time I use this machine, do I have to use useEffect hook to persist my state. Is there anything that will do this trick inside toggleMachine and I don't have to care about it in my other components.
Also, I thought I can use the service I created above in all of my components. Look, basically I just want to know if there's any way to do something every time the state changes without using useEffect hook? That's it. 
Thanks a lot btw for helping me.

**Reply by davidkpiano** [reply]
> basically I just want to know if there's any way to do something every time the state changes without using useEffect hook?

But `useEffect` _is_ the way to do this. If you're asking if there is a convenient built-in XState API for exactly this use-case, there is not.

**Reply by theprantadutta** [reply]
Ok, thanks anyway

**Reply by theprantadutta** [reply]
> > basically I just want to know if there's any way to do something every time the state changes without using useEffect hook?
> 
> But `useEffect` _is_ the way to do this. If you're asking if there is a convenient built-in XState API for exactly this use-case, there is not.

Just One more question, when should I use an interpreter then?

**Reply by Andarist** [reply]
I'm not sure if I got the intention right because I'm skimming really quickly through this - but if you want to "listen" to the existing service you can use `useActor` hook. Note that this will still provide you a state so might not cover some use cases that would like to utilize some other listeners~

**Reply by Andarist** [reply]
> Just One more question, when should I use an interpreter then?

If you want to create a live service from a machine's blueprint outside of React (which does exactly this within `useMachine`, it's just abstracted from you).

**Reply by theprantadutta** [reply]
Thanks, @davidkpiano & @Andarist Thanks for helping me out guys. The way I solved this is I used a service inside my authMachine.ts file. Then I am accessing that in any of my react components. Now I am using this same service for all of my react components. And every time my state is changed, I am getting a console log. This is exactly what I wanted.
